From e11d6e8fb9aa1babe880d9ced61088ea856bcf88 Mon Sep 17 00:00:00 2001
From: Gopi Kumar Bulusu <gopi@sankhya.com>
Date: Wed, 1 May 2024 11:12:32 +0530
Subject: [PATCH 4/4] When unwinding pc value, adjust return pc value

A call (branch and link) instruction can include a delay slot, the
value of pc stored in the link register for Microblaze architecture
is the pc value corresponding to last executed instruction (call)
in the caller. The return instruction (branch reg) includes an
offset of 8 so that when function returns execution continues from
the address at : link register + 8, as the instruction in delay slot
(link register + 4) is already executed at the time of call.

Handle this by adjusting pc value during unwind-pc.

Basically restoring code to do this that seems to have been removed
as part of a gdb patch (gdb patch #8, Xilinx Yocto 2023.2)

That patch caused hundreds of regressions in gdb testuite, including
gdb.base/advance.exp, which is now fixed.

Signed-off-by: Gopi Kumar Bulusu <gopi@sankhya.com>
Signed-off-by: Aayush Misra <aayushm@amd.com>

Upstream-Status: Pending
---
 gdb/microblaze-tdep.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/gdb/microblaze-tdep.c b/gdb/microblaze-tdep.c
index a40edf10259..aaf6873822e 100644
--- a/gdb/microblaze-tdep.c
+++ b/gdb/microblaze-tdep.c
@@ -373,6 +373,12 @@ microblaze_unwind_pc (struct gdbarch *gdbarch, const frame_info_ptr &next_frame)
   CORE_ADDR pc;
 
   pc=frame_unwind_register_unsigned (next_frame, MICROBLAZE_PC_REGNUM);
+  /* For sentinel frame, return address is actual PC.  For other frames,
+     return address is pc+8.  This is a workaround because gcc does not
+     generate correct return address in CIE.  */
+  if (frame_relative_level (next_frame) >= 0)
+    pc = pc + 8;
+  microblaze_debug ("unwind pc = 0x%x\n", (int) pc);
   return pc;
 }
 
-- 
2.47.3

